src/prefect/events/worker.py

```diff
@@ -0,0 +1,128 @@
+import asyncio
+import atexit
+import queue
+import threading
+from contextlib import asynccontextmanager, contextmanager
+from typing import Any, Dict, Optional, Type
+
+from prefect._internal.compatibility.experimental import experiment_enabled
+from prefect.settings import PREFECT_API_KEY, PREFECT_API_URL, PREFECT_CLOUD_API_URL
+
+from .clients import EventsClient, NullEventsClient, PrefectCloudEventsClient
+from .schemas import Event
+
+
+class EventsWorker:
+    def __init__(
+        self,
+        client_type: Type[EventsClient],
+        client_kwargs: Optional[Dict[str, Any]] = None,
+    ):
+        self._client_type = client_type
+        self._client_kwargs = client_kwargs or {}
+        self._client: EventsClient
+
+        self._queue: queue.SimpleQueue[Event] = queue.SimpleQueue()
+
+        self._thread = threading.Thread(
+            target=self._start_loop, name="events-worker", daemon=True
+        )
+        self._stop_event = threading.Event()
+        self._ready_event = threading.Event()
+        self._lock = threading.Lock()
+
+        atexit.register(self.stop)
+
+    def _start_loop(self):
+        """Entrypoint for the thread."""
+        loop = asyncio.new_event_loop()
+        loop.run_until_complete(self._main_loop())
+        loop.close()
+
+    async def _main_loop(self):
+        """Orchestrate the events client and emit any queued events."""
+        self._client = self._client_type(**self._client_kwargs)
+        async with self._client:
+            self._ready_event.set()
+            while not self._stop_event.is_set():
+                try:
+                    event = self._queue.get_nowait()
+                except queue.Empty:
+                    await asyncio.sleep(0.1)
```

Comment:
This is not an performant way to poll. You should probably use an `asyncio.Queue` so you can `await` a `get` call properly then you can put items in the queue like https://github.com/PrefectHQ/prefect/blob/main/src/prefect/_internal/concurrency/supervisors.py#L330-L331

Reply:
Awesome, thanks Zanie! I'll refactor that bit.

Reply:
Actually I explicitly don't want to use an async queue, the _main_ point of this whole thing is to ensure that we can communicate with the events stuff synchronously.  Is there a better way to poll a `SimpleQueue` ?

Reply:
I don't think you need to use a worker thread to `put_nowait` into an `asyncio.Queue`, it's a sync function:

```
$ python
Python 3.10.6 (main, Dec 15 2022, 11:19:56) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import asyncio
>>> q = asyncio.Queue()
>>> q.put_nowait('hi')
>>> q
<Queue at 0x7f72e95d1cc0 maxsize=0 _queue=['hi'] tasks=1>
```

Reply:
Without using a `get_nowait` and a `sleep`, how would it be possible to stop this thread?

Reply:
@madkinsz explained this to me. Essentially the idea here is that it blocks on start until something goes into the queue and then to stop it you enqueue some sentinel value (like `None`) that signals the thread to unblock and shutdown.

Reply:
Ahh yes, a good old sentinel, that works!

Reply:
@chrisguidry if you put items in the `asyncio.Queue` from outside the event loop that the queue belongs to, it won't wake up the queue when an item is placed in it because the implementation is not threadsafe. That's why you need to wrap `put_nowait` with `call_soon_threadsafe` even if it's just a sync function.

Reply:
Ohhh that makes sense, thank you for clarifying that.  I guess I've almost always been operating in a one-loop situation and not fretting too much about the context switch between sync and async.